semantics/cpp14/language/translation/decl/enum.k

```diff
@@ -0,0 +1,238 @@
+          declareEnumerators(E::CPPEnumType, Min::Int, Max::Int, Next::Int, PrevType::CPPSimpleType, ListItem(Enumerator(X::CId, NoExpression())) L::List, false, true)
+          => recoverFromError(
+          declareEnumerators(E, Min, Max, Next, PrevType, ListItem(Enumerator(X, prv(0, noTrace, type(int)))) L, false, true)
+          ))
+
+     // Second or more enumerator, fixed underlying type
+     rule declareEnumerators(E::CPPEnumType, _, _, Next::Int, _,
+               ListItem(Enumerator(_, NoExpression() => prv(Next, noTrace, underlyingType(E)))) _::List, true, false)
+
+     // Second or more enumerator, underlying type not fixed, value fits.
+     rule declareEnumerators(_, _, _, Next::Int, PrevType::CPPSimpleType,
+               ListItem(Enumerator(_, NoExpression() => prv(Next, noTrace, type(PrevType)))) _::List, false, false)
+          requires inRange(Next, type(PrevType))
+
+     // Second or more enumerator, underlype type not fixed, value does not fit to previous type but fits to some.
+     rule (.K => UNSPEC("TDE3", "Value does not fit to type of preceding enumerator")) ~> (
+          declareEnumerators(E::CPPEnumType, Min::Int, Max::Int, Next::Int, PrevType::CPPSimpleType, ListItem(Enumerator(X::CId, NoExpression())) L::List, false, false)
+          => recoverFromError(
+          declareEnumerators(E, Min, Max, Next::Int, PrevType, ListItem(Enumerator(X, prv(Next, noTrace, type(firstSuitableType(Next, Next, allIntegralTypes))))) L, false, false)
+          ))
+          requires notBool inRange(Next, type(PrevType)) andBool firstSuitableType(Next, Next, allIntegralTypes) =/=K no-type
+
+     rule (.K => ILL("TDE4", "Incremented enumerator value is not representable in any integer type")) ~>
+          declareEnumerators(_, _, _, Next::Int, PrevType::CPPSimpleType, ListItem(Enumerator(_, NoExpression())) _::List, false, false)
+          requires notBool inRange(Next, type(PrevType)) andBool firstSuitableType(Next, Next, allIntegralTypes) ==K no-type
+
+     rule declareEnumerators(_, _, _, _, _, ListItem(Enumerator(_, prv(_, _, T:CPPEnumType => underlyingType(T)))) _::List, _, _)
+     context declareEnumerators(_, _, _, _, _, ListItem(Enumerator(_, HOLE:Expr)) _::List, _, _)
+
+     // Type of an enumerator inside the declaration is enum's underlying type
+     // if it is fixed, and type of initializing expression otherwise,
+     // Hence we need to store the type of underlying expression...
+     rule declareEnumerators(T::CPPEnumType, Min::Int, Max::Int, Next::Int, _, ListItem(Enumerator(X::CId, prv(V::Int, _, t(_, _, EType::CPPSimpleType)) #as Pr::PRVal)) L::List, Fixed:Bool, First:Bool)
+          => declareEnumerator(getEnumId(T), X, Pr, Fixed, underlyingType(T))
+          ~> declareEnumerators(T, #if First #then V #else minInt(Min, V) #fi, #if First #then V #else maxInt(Max, V) #fi, V +Int 1, EType, L, Fixed, false)
```

Comment:
this rule needs a side condition declaring that EType isn't a CPPSimpleUnscopedEnumType.